engine/src/test/java/org/hibernate/validator/test/internal/engine/cascaded/NestedCascadedConstraintsTest.java

```diff
@@ -0,0 +1,150 @@
+
+		assertNumberOfViolations( constraintViolations, 0 );
+
+		constraintViolations = validator.validate( EmailAddressMap.invalidEmailAddressMap() );
+
+		assertNumberOfViolations( constraintViolations, 2 );
+		assertCorrectPropertyPaths(
+				constraintViolations,
+				"map[invalid].[1].email",
+				"map[invalid].[2].email"
+		);
+		assertCorrectConstraintTypes( constraintViolations, Email.class, Email.class );
+	}
+
+	@Test
+	public void testMultipleNestedValidWithCustomExtractor() {
+		Validator validator = Validation.byProvider( HibernateValidator.class )
+				.configure()
+				.addCascadedValueExtractor( new ReferenceValueExtractor() )
+				.buildValidatorFactory()
+				.getValidator();
+
+		Set<ConstraintViolation<CinemaEmailAddresses>> constraintViolations = validator.validate( CinemaEmailAddresses.validCinemaEmailAddresses() );
+
+		assertNumberOfViolations( constraintViolations, 0 );
+
+		constraintViolations = validator.validate( CinemaEmailAddresses.invalidCinemaEmailAddresses() );
+
+		assertNumberOfViolations( constraintViolations, 2 );
+		assertCorrectPropertyPaths(
+				constraintViolations,
+				"map[Optional[Cinema[cinema2]]].[1].email",
+				"map[Optional[Cinema[cinema2]]].[2].email"
+		);
+
+		constraintViolations = validator.validate( CinemaEmailAddresses.invalidKey() );
+
+		assertNumberOfViolations( constraintViolations, 2 );
+		assertCorrectPropertyPaths(
+				constraintViolations,
+				"map[Optional[Cinema[cinema3]]].visitor.name",
+				"map[Optional.empty].<map key>"
+		);
+	}
+
+	private static class EmailAddressMap {
+
+		private final Map<String, @Valid List<@Valid EmailAddress>> map = new HashMap<>();
+
+		private static EmailAddressMap validEmailAddressMap() {
+			List<EmailAddress> validEmailAddresses = Arrays.asList( new EmailAddress( "valid-email-1@example.com" ), new EmailAddress( "valid-email-2@example.com" ) );
+
+			EmailAddressMap emailAddressMap = new EmailAddressMap();
+			emailAddressMap.map.put( "valid", validEmailAddresses );
+
+			return emailAddressMap;
+		}
+
+		private static EmailAddressMap invalidEmailAddressMap() {
+			List<EmailAddress> validEmailAddresses = Arrays.asList( new EmailAddress( "valid-email-1@example.com" ), new EmailAddress( "valid-email-2@example.com" ) );
+			List<EmailAddress> emailAddressesContainingInvalidEmails = Arrays.asList( new EmailAddress( "valid-email-3@example.com" ),
+					new EmailAddress( "invalid-1" ), new EmailAddress( "invalid-2" ) );
+
+			EmailAddressMap emailAddressMap = new EmailAddressMap();
+			emailAddressMap.map.put( "valid", validEmailAddresses );
+			emailAddressMap.map.put( "invalid", emailAddressesContainingInvalidEmails );
+
+			return emailAddressMap;
+		}
+	}
+
+	private static class CinemaEmailAddresses {
+
+		private final Map<@NotNull @Valid Optional<@NotNull @Valid Cinema>, @Valid List<@Valid EmailAddress>> map = new HashMap<>();
```

Comment:
`Optional<@NotNull ...>` seems not sensible. Also there are some `@Valid` too much I think. This would be more realistic:

    Map<@NotNull Optional<@Valid Cinema>, List<@Valid EmailAddress>> map;

I prefer us to stick to realistic examples so we are sure we are on the right track to solve actual issues (I don't mean to sound harsh I case I do ;).